### PR TITLE
feat(mergerfs): Add extension

### LINKED ISSUE #396: Extension for mergerFS
My storage pool is largely different file systems, as a JBOD so this would help a lot if Talos had the extension so that it could be utilized.

It uses fuse underneath to my understanding and is packaged for eg debian.

https://github.com/trapexit/mergerfs

### PR DESCRIPTION
**Pull request by samip5**
This will fix #396

### CODE REVIEW

File: storage/mergerfs/mergerfs.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+name: mergerfs
+depends:
+  - configuration: true
+container:
+  entrypoint: /usr/local/bin/mergerfs
+  args:
+  - -o config=/usr/local/etc/mergerfs/config.ini
```
Comment (frezbo) [inline storage/mergerfs/mergerfs.yaml:7]: this should use an extensionserviceconfig so that the service waits on it, check the tailscale example

**Review by frezbo — COMMENTED** [review]
missing documentation, would need some docs on how to test this, so that tests maybe added in talos side

**Comment by smira** [comment]
And I think we should build from source (it shouldn't be crazy given that it's C++).

File: storage/mergerfs/mergerfs.yaml (hunk 1 above)
Comment (samip5) [reply storage/mergerfs/mergerfs.yaml:7]: I'm not at all sure, how this differs in that aspect of the tailscale example?

Comment (samip5) [reply storage/mergerfs/mergerfs.yaml:7]: @frezbo Did you not see it? https://github.com/skyssolutions/talos-extensions/blob/2d78d4eec93c7695b67e21a7ac3187a2be3e509d/storage/mergerfs/mergerfs.yaml#L3

**Comment by samip5** [comment]
I'm also little confused as to why it would try to create /var/mnt, instead of bind mounting it..

```
ID       ext-mergerfs
STATE    Failed
HEALTH   ?
EVENTS   [Failed]: Failed to create runner: mkdir /var/mnt: read-only file system (3m45s ago)
         [Preparing]: Creating service runner (3m45s ago)
         [Preparing]: Running pre state (3m46s ago)
         [Waiting]: Waiting for extension service config (4m11s ago)
         [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m13s ago)
         [Waiting]: Waiting for service "containerd" to be registered, extension service config (4m16s ago)
         [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m18s ago)
         [Starting]: Starting service (4m18s ago)
```

File: storage/mergerfs/mergerfs.yaml (hunk 1 above)
Comment (frezbo) [reply storage/mergerfs/mergerfs.yaml:7]: oh I missed that, sorry

**Comment by frezbo** [comment]
> I'm also little confused as to why it would try to create /var/mnt, instead of bind mounting it..
> 
> ```
> ID       ext-mergerfs
> STATE    Failed
> HEALTH   ?
> EVENTS   [Failed]: Failed to create runner: mkdir /var/mnt: read-only file system (3m45s ago)
>          [Preparing]: Creating service runner (3m45s ago)
>          [Preparing]: Running pre state (3m46s ago)
>          [Waiting]: Waiting for extension service config (4m11s ago)
>          [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m13s ago)
>          [Waiting]: Waiting for service "containerd" to be registered, extension service config (4m16s ago)
>          [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m18s ago)
>          [Starting]: Starting service (4m18s ago)
> ```
> 
> ```yaml
> apiVersion: v1alpha1
> kind: ExtensionServiceConfig
> name: mergerfs
> configFiles:
>   - content: |-
>       [mount]
>       path = /var/mnt/merged
>       fstype = fuse.mergerfs
>       options = defaults,allow_other,use_ino,nonempty,minfreespace=50G,category.create=mfs,moveonenospc=true,fsname=mergerFS
>       [paths]
>       source = /var/mnt/disk*
>     mountPath: /usr/local/etc/mergerfs/config.ini
> ```

you'd need a dependency on `service: cri` to wait for `/var` to be mounted

**Comment by samip5** [comment]
> > I'm also little confused as to why it would try to create /var/mnt, instead of bind mounting it..
> 
> > 
> 
> > ```
> 
> > ID       ext-mergerfs
> 
> > STATE    Failed
> 
> > HEALTH   ?
> 
> > EVENTS   [Failed]: Failed to create runner: mkdir /var/mnt: read-only file system (3m45s ago)
> 
> >          [Preparing]: Creating service runner (3m45s ago)
> 
> >          [Preparing]: Running pre state (3m46s ago)
> 
> >          [Waiting]: Waiting for extension service config (4m11s ago)
> 
> >          [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m13s ago)
> 
> >          [Waiting]: Waiting for service "containerd" to be registered, extension service config (4m16s ago)
> 
> >          [Waiting]: Waiting for service "containerd" to be "up", extension service config (4m18s ago)
> 
> >          [Starting]: Starting service (4m18s ago)
> 
> > ```
> 
> > 
> 
> > ```yaml
> 
> > apiVersion: v1alpha1
> 
> > kind: ExtensionServiceConfig
> 
> > name: mergerfs
> 
> > configFiles:
> 
> >   - content: |-
> 
> >       [mount]
> 
> >       path = /var/mnt/merged
> 
> >       fstype = fuse.mergerfs
> 
> >       options = defaults,allow_other,use_ino,nonempty,minfreespace=50G,category.create=mfs,moveonenospc=true,fsname=mergerFS
> 
> >       [paths]
> 
> >       source = /var/mnt/disk*
> 
> >     mountPath: /usr/local/etc/mergerfs/config.ini
> 
> > ```
> 
> 
> 
> you'd need a dependency on `service: cri` to wait for `/var` to be mounted

That results in other problems as cri will not come up if all user disks are not able to be mounted because they're eg non xfs. The whole point of mergerfs is to use disks that are separate and potentially different file systems. 

Related: https://github.com/siderolabs/talos/issues/9023

**Comment by samip5** [comment]
Closed due to better wait for https://github.com/siderolabs/talos/issues/8367 and this requires a lot of the filesystem tools to exist on rootfs which they currently don't.

Like ext* related utilities, and btrfs to name a few.

Also related to: https://github.com/siderolabs/talos/pull/9026

**Comment by lehmanju** [comment]
Any update on the feasability of this? The linked issues have been solved.

**Pull request closed without merging**